lib/auth/auth_with_roles.go

```diff
@@ -5455,11 +5455,29 @@ func (a *ServerWithRoles) GetAccountRecoveryCodes(ctx context.Context, req *prot
 }
 
 // GenerateCertAuthorityCRL generates an empty CRL for a CA.
+//
+// This CRL can be requested by:
+//
+//   - Windows desktop service when updating the certificate authority contents
+//     on LDAP.
+//   - Cluster administrator using "tctl auth crl --type=db" command locally
+//     on the auth server to produce revocation list used to be configured on
+//     external services such as Windows certificate store.
+//   - Remote user using "tctl auth crl --type=db" command with a remote
+//     proxy (e.g. Teleport Cloud), as long as they have permission to read
+//     certificate authorities.
 func (a *ServerWithRoles) GenerateCertAuthorityCRL(ctx context.Context, caType types.CertAuthType) ([]byte, error) {
```

Comment:
Suggesting some internal comments to help explain the logic:

```go
func (a *ServerWithRoles) GenerateCertAuthorityCRL(ctx context.Context, caType types.CertAuthType) ([]byte, error) {
	// Assume this is a user request, check if the user has permission to read CRLs.
	err := a.action(apidefaults.Namespace, types.KindCertAuthority, types.VerbReadNoSecrets)
	if err != nil {
		// An error either means the user doesn't have permission to read CRLs, or
		// this is an admin on the auth server, or the windows desktop service.
		// In any of those cases we expect to see an access denied error.
		if !trace.IsAccessDenied(err) {
			return nil, trace.Wrap(err)
		}

		// If this is an admin on the auth server (types.RoleAdmin)
		// or the windows desktop service (types.RoleWindowsDesktop),
		// allow the request, otherwise return the access denied error.
		if !a.hasBuiltinRole(types.RoleAdmin, types.RoleWindowsDesktop) {
			return nil, trace.Wrap(err)
		}
	}

	crl, err := a.authServer.GenerateCertAuthorityCRL(ctx, caType)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return crl, nil
}
```

Also note that

```go
if !a.hasBuiltinRole(types.RoleAdmin, types.RoleWindowsDesktop) {
	return nil, trace.AccessDenied("access denied")
}
```
has been changed to
```go
if !a.hasBuiltinRole(types.RoleAdmin, types.RoleWindowsDesktop) {
	return nil, trace.Wrap(err)
}
```

Reply:
Updated the code with your suggestions. Thanks!